engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateByBucketHelper.java

```diff
@@ -0,0 +1,336 @@
+
+                // no need to consider upstream removals
+                previousToShift.remove(upstream.removed());
+
+                final RowSetShiftData.Iterator sit = upstream.shifted().applyIterator();
+                while (sit.hasNext()) {
+                    sit.next();
+                    try (final RowSet subRowSet = previousToShift.subSetByKeyRange(sit.beginRange(), sit.endRange())) {
+                        if (subRowSet.isEmpty()) {
+                            continue;
+                        }
+
+                        final LongChunk<? extends Values> shiftValues =
+                                timestampColumnSource.getPrevChunk(getContext, subRowSet).asLongChunk();
+
+                        timestampSsa.applyShiftReverse(shiftValues, subRowSet.asRowKeyChunk(), sit.shiftDelta());
+                    }
+                }
+            }
+        }
+
+        // adds
+        if (restampAdditions.isNonempty()) {
+            final int size = (int) Math.min(restampAdditions.size(), 4096);
+            try (final RowSequence.Iterator it = restampAdditions.getRowSequenceIterator();
+                    final ChunkSource.GetContext context = timestampColumnSource.makeGetContext(size);
+                    final WritableLongChunk<? extends Values> ssaValues = WritableLongChunk.makeWritableChunk(size);
+                    final WritableLongChunk<OrderedRowKeys> ssaKeys = WritableLongChunk.makeWritableChunk(size)) {
+                MutableLong lastTimestamp = new MutableLong(NULL_LONG);
+                while (it.hasMore()) {
+                    RowSequence chunkRs = it.getNextRowSequenceWithLength(4096);
+
+                    // get the chunks for values and keys
+                    LongChunk<? extends Values> valuesChunk =
+                            timestampColumnSource.getChunk(context, chunkRs).asLongChunk();
+                    LongChunk<OrderedRowKeys> keysChunk = chunkRs.asRowKeyChunk();
+
+                    // push only non-null values/keys into the Ssa
+                    fillChunkWithNonNull(keysChunk, valuesChunk, ssaKeys, ssaValues, lastTimestamp);
+                    timestampSsa.insert(ssaValues, ssaKeys);
+                }
+            }
+        }
+    }
+
+    /**
+     * helper function to fill a LongChunk while skipping values that are NULL_LONG. Used to populate an SSA from a
+     * source containing null values
+     */
+    private void fillChunkWithNonNull(LongChunk<OrderedRowKeys> keysChunk, LongChunk<? extends Values> valuesChunk,
+            WritableLongChunk<OrderedRowKeys> ssaKeys, WritableLongChunk<? extends Values> ssaValues,
+            MutableLong lastTimestamp) {
+        // reset the insertion chunks
+        ssaValues.setSize(0);
+        ssaKeys.setSize(0);
+
+        // add only non-null timestamps to this Ssa
+        for (int i = 0; i < valuesChunk.size(); i++) {
+            long ts = valuesChunk.get(i);
+            if (ts < lastTimestamp.longValue()) {
```

Comment:
`lastTimestamp` is never set properly. Don't you need/want to validate intra-chunk non-descending, too?

Reply:
oops, nice catch!